### PR TITLE
Show fnsig's unit output  explicitly when there is output diff in diagnostics

### LINKED ISSUE #127263: Assigning the wrong fn type to a variable mentions 'pointer' vs 'item'
### Code

```Rust
fn bar() {}

fn main() {
    let b: fn() -> u32 = bar;
}
```

### Current output

```Shell
error[E0308]: mismatched types
 --> src/main.rs:4:26
  |
4 |     let b: fn() -> u32 = bar;
  |            -----------   ^^^ expected fn pointer, found fn item
  |            |
  |            expected due to this
  |
  = note: expected fn pointer `fn() -> u32`
                found fn item `fn() {bar}`
```

### Desired output

```Shell
error[E0308]: mismatched types
 --> src/main.rs:4:26
  |
4 |     let b: fn() -> u32 = bar;
  |            -----------   ^^^ `fn` return types do not match, expected `fn() -> u32`, found `fn() -> ()`
  |            |
  |            expected due to this
```

### Rationale and extra context

The current version is not _wrong_, but it is a bit confusing when one does not know the way rustc handles `fn` stuff.

I think that leading with what is not matching up would make it easier to see one's error. Especially in context for example of RA, where in-line errors only show the 'main' message.

### Other cases

_No response_

### Rust Version

```Shell
Playground, Stable 1.79.0
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by chenyukang**
Fixes #127263

### CODE REVIEW

File: compiler/rustc_infer/src/infer/error_reporting/mod.rs [hunk 1]
```diff
@@ -1168,14 +1168,21 @@ impl<'a, 'tcx> TypeErrCtxt<'a, 'tcx> {
         let output1 = sig1.output();
         let output2 = sig2.output();
         let (x1, x2) = self.cmp(output1, output2);
-        if !output1.is_unit() {
+        let only_output_diff = !lifetime_diff
```
Comment (oli-obk) [inline compiler/rustc_infer/src/infer/error_reporting/mod.rs:1171]: Why not just show it whenever the outputs differ?

File: tests/ui/compare-method/bad-self-type.stderr [hunk 2]
```diff
@@ -39,7 +39,7 @@ note: type in trait
 LL |     fn bar(self) -> Option<()>;
    |                     ^^^^^^^^^^
    = note: expected signature `fn(MyFuture) -> Option<()>`
-              found signature `fn(MyFuture)`
+              found signature `fn(MyFuture) -> ()`
```
Comment (oli-obk) [inline tests/ui/compare-method/bad-self-type.stderr:42]: Not sure how I feel about those... but seems ok

**Comment by oli-obk** [comment]
Needs a rebase

File: compiler/rustc_infer/src/infer/error_reporting/mod.rs (hunk 1 above)
Comment (chenyukang) [reply compiler/rustc_infer/src/infer/error_reporting/mod.rs]: yeah, seems the result not bad:
https://github.com/rust-lang/rust/pull/127417/commits/f46c4129e0310b4fb6a75dbe28c2a86e00c9be27

**Comment by oli-obk** [comment]
bors r plus


@bors rollup

**Pull request merged**